Fix up tests and fix up splitio-adapter

tdeekens opened the pull request:
💯 Thanks again for your contribution. This is my first attempt to "fix up" the remaining bits of the adapter. I hope it makes sense I will inline some explanations in the code.

tdeekens committed:
test(splitio/adapter): fix testing of splitio library

tdeekens committed:
test(splitio/adapter): the normalization of flag name and value

tdeekens committed:
refactor(splitio/adapter): rename client initializationw fn

tdeekens committed:
refactor(splitio/adapter): rename clientSideId to authorizationKey

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.js:
```diff
@@ -9,27 +9,30 @@ const adapterState = {
-const normalizeFlag = (flagName, flagValue) => {
-  let value;
+export const normalizeFlag = (flagName, flagValue) => {
+  let normalizeFlagValue;
```
This is just a bit of a naming dance to keep things clearer. Hopefully.

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.js:
```diff
@@ -86,14 +90,14 @@ const subscribe = ({ onFlagsStateChange, onStatusStateChange }) =>
 
 const configure = ({
-  clientSideId,
+  authorizationKey,
```
I felt we should rename this to what it actually is in a splitio context.

Kerumen replied:
Yup, I named like this first but I wanted to expose the same API as LD. `authorizationKey` fit better.

tdeekens replied:
Yeah I think it can really and should depend on the adapter. Also so that it reads nice on the configuration side

```jsx
<ConfigureFlopflip adapter={splitioAdapter} adapterArgs={{ authorizationKey }} />
```

Note that the `adapterArgs` shoudl always be a memoized thing so the reference does not change on each call to render.

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.js:
```diff
@@ -86,14 +90,14 @@ const subscribe = ({ onFlagsStateChange, onStatusStateChange }) =>
   adapterState.user = ensureUser(user);
-  const { client, manager } = initializeUserContext(
-    clientSideId,
+  const { client, manager } = initializeClient(
```
This also is more of client inititialization to me to get the `client` and `manager`.

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.spec.js:
```diff
@@ -1,32 +1,12 @@
+import adapter, {
+  camelCaseFlags,
+  createAnonymousUserKey,
+  normalizeFlag,
```
...so we can test this a bit below.

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.js:
```diff
@@ -50,13 +53,14 @@ const ensureUser = user => ({
 });
 
-const initializeUserContext = (clientSideId, user) => {
+const initializeClient = (authorizationKey, user) => {
```
This notion of user is very interesting to me. I always felt each adapter/provider of flags will have a user notion and would provide custom properties to be fed in for targeting. For splitio, according to their [docs](https://docs.split.io/docs/javascript-sdk-overview) you only pass in a key.

Kerumen replied:
Yes absolutely!

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.spec.js:
```diff
@@ -100,15 +88,18 @@ describe('when configuring', () => {
       factory = {
         client: jest.fn(() => ({
-          on: jest.fn((_, cb) => cb()),
+          on: onStub,
```
Otherwise the the `splitio()` call will generate through `client()` a new `on` stub each time. Especially down in the previous expectation of `client().on` which then has a zero'ed call count.

Kerumen replied:
That's what I didn't succeed to do. Nice spot :)

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.spec.js:
```diff
@@ -154,45 +145,58 @@ describe('when configuring', () => {
         onStatusStateChange = jest.fn();
         onFlagsStateChange = jest.fn();
+        namesStub = jest.fn(() => names);
+        getTreatmentsStub = jest.fn(() => flags);
```
Same as above applies for these two.

tdeekens commented on packages/splitio-adapter/modules/adapter/adapter.spec.js:
```diff
@@ -235,3 +239,54 @@ describe('create anonymous user', () => {
+describe('normalizeFlag', () => {
+  const flagName = 'fooFlag';
+
+  describe('with `flagValue` being `null`', () => {
```
Just wanted to make sure we get good coverage on these.

tdeekens reviewed (commented):
See some explanation attached. If you merge this into your PR it should update your PR against `flopflip` itself and we can then take it from there. Do you/we have any change to "integration" test the adapter to at least try it out? Did you maybe already do that?

Kerumen commented:
@tdeekens Thanks a lot to fix up my tests and comment the code. Everything is a lot clearer now. I'm gonna merge it.

> Do you/we have any change to "integration" test the adapter to at least try it out?

I didn't understand this part tho.. 😶

Kerumen commented on packages/splitio-adapter/modules/adapter/adapter.js:
```diff
@@ -50,13 +53,14 @@ const ensureUser = user => ({
   const factory = splitio({
     core: {
-      authorizationKey: clientSideId,
+      authorizationKey: authorizationKey,
```
You can remove the value part. It's ES6 😉 
```

Kerumen commented on packages/splitio-adapter/modules/adapter/adapter.js:
```diff
@@ -50,13 +53,14 @@ const ensureUser = user => ({
   const factory = splitio({
     core: {
-      authorizationKey: clientSideId,
+      authorizationKey: authorizationKey,
```
You can remove the value part. It's ES6 😉 
```
authorizationKey,
```

tdeekens replied:
True that. Nice spot will do that later.

tdeekens commented:
By integration tests I meant, and this will become more important with the growth of the project, how I/we can actually test that the adapters work. Up until now I always used the demo to quickly spin up things and test them. That would be nice to automate. So I unit test all adapters, React components et al but I also integration test a setup. So I configure flopflip with each adapter switch a flag and see if things on the "page" actually change.

Kerumen commented:
Okay I see. FWIW, I use this version of the adapter in production since 2 days (with a split.io account) and it works really well.

tdeekens commented:
Wow, super nice it works for you that makes me happy. Means that the concept of adapters really kind of works and was a right move for the library. Thanks again for contributing back!

tdeekens committed:
refactor(splitio/adapter): remove uncalled for key value passing

tdeekens committed:
docs(splitio/adapter): add to readme dependency status

tdeekens commented:
Fixed up your request. Thanks for the wink. I also added the adapter to the readme.

After you merge this into your branch please rebase the branch on top of `master` or merge master in. I think we will have a problem bundling the adapter due to some of splitio's dependencies. At least I found out yesterday. But I will have a look then.
